Change `ChunkReader.Chunk()` to `ChunkOrIterable()`

fionaliao opened the pull request:
<!--
    Don't forget!

    - Please sign CNCF's Developer Certificate of Origin and sign-off your commits by adding the -s / --signoff flag to `git commit`. See https://github.com/apps/dco for more information.

    - If the PR adds or changes a behaviour or fixes a bug of an exported API it would need a unit/e2e test.

    - Where possible use only exported APIs for tests to simplify the review and make it as close as possible to an actual library usage.

    - No tests are needed for internal implementation changes.

    - Performance improvements would need a benchmark test to prove it.

    - All exposed objects should have a comment.

    - All comments should start with a capital letter and end with a full stop.
 -->

This change is related to OOO native histogram support: https://github.com/prometheus/prometheus/issues/11220.

## Problem
`oooMergedChunk` currently implements the `Chunk` interface (i.e. it's considered a chunk). A chunk cannot have a histogram counter reset in the middle of it. With OOO native histograms, `oooMergedChunk` can break this expectation. This is because  `oooMergedChunk` can contain several OOO native histograms chunks. These chunks can start with counter resets. If you merge the chunks together, then there could be counter reset in the middle of the `oooMergedChunk`.

The reason that `oooMergedChunk` implements the `Chunk` interface is that the querier, specifically the `populateWithDelChunkSeriesIterator`, uses a `ChunkReader` with the `Chunk(chunks.Meta)` method. This method expects only a single `Chunk` to be returned. A single `chunks.Meta` can actually refer to multiple chunks for OOO (`memSeries.oooMergedChunk` finds all of the chunks within a given `mint` and `maxt` for a given `chunks.Meta`).

This is problematic as during compaction (which uses `populateWithDelChunkSeriesIterator`), an `oooMergedChunk` could be written as a single chunk in a block, even if it has a counter reset in the middle. Also if any tombstones are applied, the iterator will iterate through all of the samples in the `oooMergedChunk` and attempt to append it to a single chunk - this will error if there's a counter reset in the middle of `oooMergedChunk`.

## Proposal
My proposal is to change the `ChunkReader.Chunk()` to `ChunkOrIterator()` - so it'll return either a single chunk or an iterator. The caller will have to convert the samples in the iterator into chunks. This could result in multiple chunks. 

The other changes required for OOO native histograms are currently being done in https://github.com/grafana/mimir-prometheus/tree/cedwards/ooo-native-histograms. This change has been separated into its own PR so the proper OOO PR will be smaller.

## Further advantages of the refactoring
- This is also useful for the case when there are multiple different encodings in a single `oooMergedChunk` - we will create a new chunk if the encoding for the next sample is different from the previous one.
- Improves performance of compacting OOO chunks in the head. Currently `Chunk.Bytes()` is called multiple times during compaction. This is an expensive option for `oooMergedChunks` as that iterates through all of the samples in its underlying OOO chunks every time bytes is called. Now the `populateWithDelChunkSeriesIterator` will use the iterator returned from `ChunkOrIterator()` to create proper chunks which have more efficient `Bytes()` functions.  See benchmark comment below for stats.
- I was able to remove some methods from `oooMergedChunks` and `boundedChunk` that were only implemented to satisfy the `Chunk` interface. Now these struct only need to satisfy the `chunkenc.Iterable` interface that has a single `Iterator()` method.

## Behaviour change
- If there are multiple encodings inside a chunk, `populateWithDelChunkSeriesIterator` will create multiple chunks rather than erroring, even for in-order chunks. I think this is an acceptable change. I don't think it's currently possible to hit this case with in-order chunks, or with float-only OOO support.

## Additional refactoring left for later
- There is additional refactoring I'd like to consider (e.g. moving `Chunks` from `chunkenc.Meta` since with OOO, it's not guaranteed a single `chunkenc.Meta` will map to a single `Chunk`) but this PR is big enough that I'll leave that for another PR once this one is done.
- Also see PR comments about removing appendOnly parameter (https://github.com/prometheus/prometheus/pull/13123#pullrequestreview-1731400121, https://github.com/prometheus/prometheus/pull/13123#issuecomment-1814128119)

## Discarded solutions
- Returning a chunk slice from Chunk(chunks.Meta). The samples would be read once within Chunk(chunks.Meta), but we might need to read and copy the samples again if any tombstones are applied, or if the per-sample iterator was used (populateWithDelSeriesIterator).
- Keep returning a "Chunk" for oooMergedChunks, but in `populateWithDelChunkSeriesIterator` actually split that into multiple chunks. It looks less messy and less complicated for callers to handle if there's only a single thing returned from the `Chunk()` function but we could easily introduce errors later on if the `Chunk()` method was used elsewhere and we didn't do the oooMergedChunks check.
- Always return an iterable. This was discarded because if a single chunk is already available and doesn't need modifications, it's not worth reading the samples and copying to a new chunk that will be exactly the same.

krajorama reviewed (commented):
Started review, already loving the fact that Appender.AppendHistogram doesn't need the "appendOnly" bool flag. I've added it out of necessity, but glad to see it go.

krajorama commented on tsdb/querier.go:
```diff
@@ -817,107 +842,173 @@ func (p *populateWithDelSeriesIterator) Err() error {
+	chunksFromDelIterIdx int
+
+	// currMetaWithChunk is current meta with its chunk field set. This meta
+	// is guaranteed to map to a single chunk. This differs from currMeta as
+	// that could refer to multiple chunks.
```
```suggestion
	// is guaranteed to map to a single chunk. This differs from 
	// populateWithDelGenericSeriesIterator.currMeta as that
	// could refer to multiple chunks.
```

krajorama commented on tsdb/querier.go:
```diff
@@ -817,107 +842,173 @@ func (p *populateWithDelSeriesIterator) Err() error {
+		// Check if the encoding has changed (i.e. we need to create a new
+		// chunk as chunks can't have multiple encoding types).
+		// For the first sample, the following condition will always be true as
+		// EncNone != EncXOR | EncHistogram | EncFloatHistogram.
```
```suggestion
		// ValNoneNone != ValFloat | ValHistogram | ValFloatHistogram.
```

krajorama reviewed (commented):
I think it's worth keeping the "error" returned from Chunk.AppendHistogram, Chunk.AppendFloatHistogram since that can also catch corrupt chunks from the disk. And in general having error returned isn't a bad idea.

But we can get rid of the "appendOnly" flag, since we are recoding individual chunks and any subset of a chunk should be possible to recode cleanly.

fionaliao commented:
> already loving the fact that Appender.AppendHistogram doesn't need the "appendOnly" bool flag. I've added it out of necessity, but glad to see it go.

@krajorama I decided to remove the appendOnly change from this PR so that there are fewer changes to review. I'll open another PR for that after this one is closed.

fionaliao commented:
## Benchmarking

I ran the compaction benchmarks on my computer (`go test -run=XXX -bench=BenchmarkCompaction -count=5 -timeout=50m`) since this change affects that. I added some simple OOO compaction tests in https://github.com/prometheus/prometheus/pull/13123/commits/bbc0d09a2452b9513907ae95dfc22d01fe3cadcc.

- There is a slight memory increase for some of the non-ooo compaction cases (up to +1.42%).
- An increase in latency for three of the non-ooo compaction cases (up to +9.00%) but the latency seems to vary a lot between each run so not sure how significant that is. 
- The ooo compaction speed and latency has improved as expected (since we're not iterating through the OOO samples multiple times anymore).

I think the reason for the increase in non-ooo compaction cases is because `populateWithDelChunkSeriesIterator` is appending chunks created from `currDelIter` to a slice, instead of just setting the chunk as a field in the iterator struct. The slice is necessary because in the case of ooo native histograms, it's possible to end up with several chunks when reading `currDelIter`. But the other cases (in order chunks, ooo floats) are more likely to happen. I'll experiment with reverting to setting the `populateWithDelChunkSeriesIterator.currMetaWithChunk` field instead of appending to a struct if only a single chunk is created  from `currDelIter`.

```
benchstat 76078d7a5-v2.txt d3086c44e-v2.txt
name                                                                             old time/op    new time/op    delta
Compaction/type=normal,blocks=4,series=10000,samplesPerSeriesPerBlock=101-16        401ms ± 3%     399ms ± 7%     ~     (p=1.000 n=5+5)
Compaction/type=normal,blocks=4,series=10000,samplesPerSeriesPerBlock=1001-16       863ms ± 5%     940ms ±15%   +9.00%  (p=0.016 n=5+5)
Compaction/type=normal,blocks=4,series=10000,samplesPerSeriesPerBlock=2001-16       1.75s ±20%     1.56s ± 8%     ~     (p=0.421 n=5+5)
Compaction/type=normal,blocks=4,series=10000,samplesPerSeriesPerBlock=5001-16       5.79s ±22%     5.08s ±19%     ~     (p=0.222 n=5+5)
Compaction/type=vertical,blocks=4,series=10000,samplesPerSeriesPerBlock=101-16      1.05s ± 5%     1.02s ±10%     ~     (p=0.548 n=5+5)
Compaction/type=vertical,blocks=4,series=10000,samplesPerSeriesPerBlock=1001-16     4.40s ± 7%     4.43s ± 6%     ~     (p=1.000 n=5+5)
Compaction/type=vertical,blocks=4,series=10000,samplesPerSeriesPerBlock=2001-16     7.89s ± 5%     7.64s ± 2%     ~     (p=0.056 n=5+5)
Compaction/type=vertical,blocks=4,series=10000,samplesPerSeriesPerBlock=5001-16     19.1s ± 3%     19.1s ± 3%     ~     (p=1.000 n=5+5)
CompactionFromHead/labelnames=1,labelvalues=100000-16                               665ms ± 3%     652ms ± 6%     ~     (p=0.310 n=5+5)
CompactionFromHead/labelnames=10,labelvalues=10000-16                               655ms ± 5%     654ms ± 4%     ~     (p=0.690 n=5+5)
CompactionFromHead/labelnames=100,labelvalues=1000-16                               673ms ± 2%     700ms ± 2%   +4.02%  (p=0.016 n=5+5)
CompactionFromHead/labelnames=1000,labelvalues=100-16                               684ms ± 4%     693ms ± 1%     ~     (p=0.730 n=5+4)
CompactionFromHead/labelnames=10000,labelvalues=10-16                               752ms ± 4%     795ms ± 2%   +5.68%  (p=0.008 n=5+5)
CompactionFromOOOHead/labelnames=1,labelvalues=100000-16                            5.07s ± 2%     2.14s ± 3%  -57.76%  (p=0.008 n=5+5)
CompactionFromOOOHead/labelnames=10,labelvalues=10000-16                            5.13s ± 2%     2.16s ± 3%  -57.97%  (p=0.008 n=5+5)
CompactionFromOOOHead/labelnames=100,labelvalues=1000-16                            5.19s ± 2%     2.18s ± 3%  -57.99%  (p=0.008 n=5+5)
CompactionFromOOOHead/labelnames=1000,labelvalues=100-16                            5.10s ± 1%     2.17s ± 4%  -57.39%  (p=0.008 n=5+5)
CompactionFromOOOHead/labelnames=10000,labelvalues=10-16                            5.21s ± 2%     2.25s ± 4%  -56.84%  (p=0.008 n=5+5)

name                                                                             old alloc/op   new alloc/op   delta
Compaction/type=normal,blocks=4,series=10000,samplesPerSeriesPerBlock=101-16       97.3MB ± 0%    98.5MB ± 0%   +1.28%  (p=0.008 n=5+5)
Compaction/type=normal,blocks=4,series=10000,samplesPerSeriesPerBlock=1001-16       136MB ± 0%     137MB ± 1%   +1.42%  (p=0.008 n=5+5)
Compaction/type=normal,blocks=4,series=10000,samplesPerSeriesPerBlock=2001-16       178MB ± 1%     179MB ± 1%     ~     (p=0.548 n=5+5)
Compaction/type=normal,blocks=4,series=10000,samplesPerSeriesPerBlock=5001-16       299MB ± 4%     300MB ± 4%     ~     (p=0.421 n=5+5)
Compaction/type=vertical,blocks=4,series=10000,samplesPerSeriesPerBlock=101-16      166MB ± 0%     167MB ± 0%   +0.75%  (p=0.008 n=5+5)
Compaction/type=vertical,blocks=4,series=10000,samplesPerSeriesPerBlock=1001-16     473MB ± 0%     476MB ± 0%   +0.48%  (p=0.008 n=5+5)
Compaction/type=vertical,blocks=4,series=10000,samplesPerSeriesPerBlock=2001-16     767MB ± 0%     768MB ± 0%     ~     (p=0.841 n=5+5)
Compaction/type=vertical,blocks=4,series=10000,samplesPerSeriesPerBlock=5001-16    1.73GB ± 0%    1.74GB ± 0%     ~     (p=0.056 n=5+5)
CompactionFromHead/labelnames=1,labelvalues=100000-16                               136MB ± 0%     136MB ± 0%     ~     (p=1.000 n=5+5)
CompactionFromHead/labelnames=10,labelvalues=10000-16                               129MB ± 0%     129MB ± 0%   -0.12%  (p=0.032 n=5+5)
CompactionFromHead/labelnames=100,labelvalues=1000-16                               132MB ± 0%     132MB ± 0%     ~     (p=0.690 n=5+5)
CompactionFromHead/labelnames=1000,labelvalues=100-16                               136MB ± 0%     136MB ± 0%     ~     (p=1.000 n=5+5)
CompactionFromHead/labelnames=10000,labelvalues=10-16                               132MB ± 0%     133MB ± 0%     ~     (p=0.548 n=5+5)
CompactionFromOOOHead/labelnames=1,labelvalues=100000-16                           1.46GB ± 0%    0.52GB ± 0%  -64.42%  (p=0.016 n=5+4)
CompactionFromOOOHead/labelnames=10,labelvalues=10000-16                           1.45GB ± 0%    0.51GB ± 0%  -64.64%  (p=0.008 n=5+5)
CompactionFromOOOHead/labelnames=100,labelvalues=1000-16                           1.45GB ± 0%    0.51GB ± 0%  -64.52%  (p=0.008 n=5+5)
CompactionFromOOOHead/labelnames=1000,labelvalues=100-16                           1.46GB ± 0%    0.52GB ± 0%  -64.38%  (p=0.008 n=5+5)
CompactionFromOOOHead/labelnames=10000,labelvalues=10-16                           1.45GB ± 0%    0.52GB ± 0%  -64.50%  (p=0.016 n=5+4)

name                                                                             old allocs/op  new allocs/op  delta
Compaction/type=normal,blocks=4,series=10000,samplesPerSeriesPerBlock=101-16         890k ± 0%      890k ± 0%     ~     (p=0.937 n=5+5)
Compaction/type=normal,blocks=4,series=10000,samplesPerSeriesPerBlock=1001-16       1.21M ± 0%     1.21M ± 0%     ~     (p=0.222 n=5+5)
Compaction/type=normal,blocks=4,series=10000,samplesPerSeriesPerBlock=2001-16       1.53M ± 0%     1.53M ± 0%     ~     (p=0.143 n=5+5)
Compaction/type=normal,blocks=4,series=10000,samplesPerSeriesPerBlock=5001-16       2.53M ± 0%     2.53M ± 0%     ~     (p=0.460 n=5+5)
Compaction/type=vertical,blocks=4,series=10000,samplesPerSeriesPerBlock=101-16      1.35M ± 0%     1.35M ± 0%     ~     (p=0.421 n=5+5)
Compaction/type=vertical,blocks=4,series=10000,samplesPerSeriesPerBlock=1001-16     3.97M ± 0%     3.97M ± 0%     ~     (p=0.222 n=5+5)
Compaction/type=vertical,blocks=4,series=10000,samplesPerSeriesPerBlock=2001-16     6.37M ± 0%     6.37M ± 0%     ~     (p=0.548 n=5+5)
Compaction/type=vertical,blocks=4,series=10000,samplesPerSeriesPerBlock=5001-16     14.3M ± 0%     14.3M ± 0%     ~     (p=0.841 n=5+5)
CompactionFromHead/labelnames=1,labelvalues=100000-16                               1.12M ± 0%     1.12M ± 0%     ~     (p=0.690 n=5+5)
CompactionFromHead/labelnames=10,labelvalues=10000-16                               1.19M ± 0%     1.19M ± 0%     ~     (p=1.000 n=5+5)
CompactionFromHead/labelnames=100,labelvalues=1000-16                               1.20M ± 0%     1.20M ± 0%     ~     (p=0.421 n=5+5)
CompactionFromHead/labelnames=1000,labelvalues=100-16                               1.22M ± 0%     1.22M ± 0%     ~     (p=0.310 n=5+5)
CompactionFromHead/labelnames=10000,labelvalues=10-16                               1.27M ± 0%     1.27M ± 0%     ~     (p=0.310 n=5+5)
CompactionFromOOOHead/labelnames=1,labelvalues=100000-16                            17.8M ± 0%      5.4M ± 0%  -69.79%  (p=0.008 n=5+5)
CompactionFromOOOHead/labelnames=10,labelvalues=10000-16                            17.8M ± 0%      5.4M ± 0%  -69.79%  (p=0.008 n=5+5)
CompactionFromOOOHead/labelnames=100,labelvalues=1000-16                            17.8M ± 0%      5.4M ± 0%  -69.78%  (p=0.008 n=5+5)
CompactionFromOOOHead/labelnames=1000,labelvalues=100-16                            17.8M ± 0%      5.4M ± 0%  -69.74%  (p=0.008 n=5+5)
CompactionFromOOOHead/labelnames=10000,labelvalues=10-16                            17.8M ± 0%      5.4M ± 0%  -69.52%  (p=0.008 n=5+5)
```

jesusvazquez commented on tsdb/head_read_test.go:
```diff
@@ -129,18 +129,7 @@ func TestBoundedChunk(t *testing.T) {
-				}
-			}
-			require.Equal(t, expChunk.Bytes(), chunk.Bytes())
+			chunk := boundedIterable{tc.inputChunk, tc.inputMinT, tc.inputMaxT}
```
nit s/chunk/iterable

fionaliao replied:
Updated in https://github.com/prometheus/prometheus/pull/13123/commits/c004db1558f20f6f8824529b712764c398f13bad

jesusvazquez reviewed (commented):
Awesome work with the issue description and the work. 

I am aware that this is a hard problem due to the constraint of the interface assuming only a Chunk could be returned and out of order breaking this premise. From the options you've worked on ChunkOrIterable is easy to understand and also seems not intrusive in the in-order code paths.

I'm going to run a benchmark on the PR to see if there are any regressions but my initial impression is that we are almost ready 👍 

Nice job, OOO for native histograms is one step closer :thisclose:

jesusvazquez commented:
/prombench main

jesusvazquez commented:
/prombench cancel

fionaliao commented:
From prombench, seeing more memory used in general and compaction is slower.

![image](https://github.com/prometheus/prometheus/assets/15159201/4f525958-1962-4f7d-b732-f77b74151e26)

![image](https://github.com/prometheus/prometheus/assets/15159201/478b4ea7-2c17-4a79-8723-2de17939e75a)

I'll experiment more with the slice used in `populateWithDelChunkSeriesIterator` (see last section of https://github.com/prometheus/prometheus/pull/13123#issuecomment-1814346958) to see if that helps. Also one alternative would be to have separate chunk iterators for ooo and in order chunks, but that would take a lot more refactoring.

jesusvazquez commented:
/prombench main

fionaliao commented:
Made some optimisation attempts in https://github.com/prometheus/prometheus/pull/13123/commits/f4a4f9fb720e7a7a11b54582f59b2a6ffd12ae50 and https://github.com/prometheus/prometheus/pull/13123/commits/61e2b8695a14215d39ce09ebf65191e0d1a77c37. From local benchmarks, not sure if that helped with anything. However, we noticed that this PR hadn't been rebased recently, and there were changes like https://github.com/prometheus/prometheus/commit/f13bc1a5c964504dc2953c6dc8420a70874e0d2e that might have affected the comparison. I've reverted my optimisation attempt and rebased, and prombench is looking a lot better.

From the increased memory usage in the local compaction benchmarks (https://github.com/prometheus/prometheus/pull/13123#issuecomment-1814346958), those are because of the new fields that were added to the `populateWithDelChunkSeriesIterator` struct (`chunksFromDelIter`, `chunksFromDelIterIdx`). Even if the fields are not touched, the memory is still increased. This doesn't seem to affect memory usage when running the prombench benchmarks.

I've just added a couple of new commits. 
https://github.com/prometheus/prometheus/pull/13123/commits/488b88f32042357567504d20cdbdd0760a1bda14 skips resetting the chunks slice in `populateWithDelChunkSeriesIterator.Next()` if we've not been reading from the deletion iterator as it's unnecessary. That might be a slight improvement for the in-order compaction path - for in-order chunk metas, `populateWithDelGenericSeriesIterator.next()` will return a single chunk rather than a deletion iterator unless there are tombstones for deleted data. 
https://github.com/prometheus/prometheus/pull/13123/commits/e7c2de2276eed2f583c4bb284c42719a8661efd1 makes sure the chunks slice is reset when `reset()` is called.

jesusvazquez commented:
/prombench cancel

jesusvazquez commented:
/prombench main

fionaliao commented:
In https://github.com/prometheus/prometheus/pull/13123/commits/9bc41a17447a2e9eb9e81f159e12c4981a111dbc, I split the reading of currDelIter into two functions - one to be called when a single chunk is returned from the chunk reader (i.e. for in-order cases), and the other to be called when an iterable is returned from the chunk reader (i.e. for ooo cases). This was done to optimise the in-order case since you know there's only going to be at most one chunk created (e.g. if there's only one chunk you don't need to add chunks created from currDelter to a slice, you can just set the chunk in currMetaWithChunk directly). This seems to have helped with compaction latency in prombench which now looks similar for this PR vs main.

Due to this change, we're not going to be able to get rid of the `appendOnly` parameter in`AppendHistogram`, since it's still being used for the single chunk case (cc @krajorama).

fionaliao commented:
There is a bug I noticed with the chunks iterator, where if currDelIter doesn't return any samples for a meta, then we don't read the samples from the remaining metas. See this test which currently fails: https://github.com/prometheus/prometheus/commit/e2156bc8e12589aeee9f607d2c1dcd57a012487f. This bug existed before my PR and my PR doesn't fix that - once this is merged, I'll make another PR for that.

bboreham commented:
/prombench cancel

jesusvazquez reviewed (approved):
Nice work @fionaliao 💪
